### PR TITLE
Allow importing certificates from authenticode assemblies

### PR DESCRIPTION
**Pull request by minego**
This change allows a number of games, such as DARK, to start successfully. The games use X509Certificate.CreateFromSignedFile() to load the certificate from their assemblies in order to verify them.

### CODE REVIEW

**Comment by marek-safar** [comment]
approve

**Comment by marek-safar** [comment]
@baulig could you please review the PR

**Comment by baulig** [comment]
I have a few cosmetic suggestions and we need to get it to build, I'll have a look at that now.

File: mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeDeformatter.cs [hunk 1]
```diff
@@ -74,12 +75,35 @@ public AuthenticodeDeformatter (string fileName) : this ()
...
+				filename = value;
+				try {
+					CheckSignature ();
+				}
+				catch (SecurityException) {
+					throw;
+				}
+				catch (Exception) {
+					reason = 1;
+				}
```
Comment (baulig) [inline mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeDeformatter.cs]: This should be

```
				try {
					CheckSignature ();
				} catch (SecurityException) {
					throw;
				} catch {
					reason = 1;
				}
```

**Comment by baulig** [comment]
To make it look a bit more pretty:

```
diff --git a/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeBase.cs b/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeBase.cs
index 814761162cb..0ed1f32a08a 100644
--- a/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeBase.cs
+++ b/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeBase.cs
@@ -175,7 +175,7 @@ namespace Mono.Security.Authenticode {
                {
                        if (fs != null)
                                Close ();
-                       fs = new MemoryStream(rawdata, false);
+                       fs = new MemoryStream (rawdata, false);
                        blockNo = 0;
                }
 
diff --git a/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeDeformatter.cs b/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeDeformatter.cs
index 156a6a3e3c6..6306f1c2377 100644
--- a/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeDeformatter.cs
+++ b/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeDeformatter.cs
@@ -87,11 +87,9 @@ namespace Mono.Security.Authenticode {
                                filename = value;
                                try {
                                        CheckSignature ();
-                               }
-                               catch (SecurityException) {
+                               } catch (SecurityException) {
                                        throw;
-                               }
-                               catch (Exception) {
+                               } catch {
                                        reason = 1;
                                }
                        }
@@ -104,11 +102,9 @@ namespace Mono.Security.Authenticode {
                                rawdata = value;
                                try {
                                        CheckSignature ();
-                               }
-                               catch (SecurityException) {
+                               } catch (SecurityException) {
                                        throw;
-                               }
-                               catch (Exception) {
```

**Comment by baulig** [comment]
We don't build `Mono.Security` on `monotouch_watch`, so to make it build we need the following:

```
diff --git a/mcs/class/System/Mono/X509PalImpl.cs b/mcs/class/System/Mono/X509PalImpl.cs
index bdfaca686d4..c9b07e2478e 100644
--- a/mcs/class/System/Mono/X509PalImpl.cs
+++ b/mcs/class/System/Mono/X509PalImpl.cs
@@ -29,8 +29,10 @@ using MonoSecurity::Mono.Security;
 using MonoSecurity::Mono.Security.Authenticode;
 #else
 using Mono.Security;
+#if !MONOTOUCH_WATCH
 using Mono.Security.Authenticode;
 #endif
+#endif
 
 using System;
 using System.IO;
@@ -128,6 +130,9 @@ namespace Mono
                                        return X509ContentType.Cert;
                        }
 
+#if MONOTOUCH_WATCH
+                       return X509ContentType.Unknown;
+#else
                        try {
                                AuthenticodeDeformatter ad = new AuthenticodeDeformatter (rawData);
 
@@ -135,8 +140,7 @@ namespace Mono
                        } catch {
                                return X509ContentType.Unknown;
                        }
-
-                       return X509ContentType.Unknown;
+#endif
                }
 
                public X509ContentType GetCertContentType (string fileName)
diff --git a/mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2ImplMono.cs b/mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2ImplMono.cs
index ee02f661f20..17eb748db34 100644
--- a/mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2ImplMono.cs
+++ b/mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2ImplMono.cs
@@ -38,7 +38,9 @@ using MX = MonoSecurity::Mono.Security.X509;
 #else
 using Mono.Security;
 using Mono.Security.Cryptography;
+#if !MONOTOUCH_WATCH
 using Mono.Security.Authenticode;
+#endif
 using MX = Mono.Security.X509;
 #endif
 
@@ -99,10 +101,12 @@ namespace System.Security.Cryptography.X509Certificates
                                _cert = new MX.X509Certificate (rawData);
                                break;
 
+#if !MONOTOUCH_WATCH
                        case X509ContentType.Authenticode:
                                AuthenticodeDeformatter ad = new AuthenticodeDeformatter (rawData);
                                _cert = ad.SigningCertificate;
                                break;
+#endif
 
                        default:
                                string msg = Locale.GetText ("Unable to decode certificate.");
```

You did not check this "allow edits from maintainers" option in your PR (https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork), so I cannot directly commit to it.

**Comment by baulig** [comment]
Once those changes are applied, it looks good to me :-)

Thank you so much for your Pull Request!

**Comment by minego** [comment]
> You did not check this "allow edits from maintainers" option in your PR (https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork), so I cannot directly commit to it.

Strange, that checkbox is checked... I don't know why it isn't letting you edit, but I am happy to apply your suggested changes.

Thank you for your feedback and review. I'll make those changes right now.

**Comment by minego** [comment]
> Once those changes are applied, it looks good to me :-)
> 
> Thank you so much for your Pull Request!

Thank you very much for the feedback and the review. The changes should be applied. Let me know if you'd like any other changes.

**Comment by baulig** [comment]
There is a test failure:

https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx/23099/testReport/junit/(root)/SoftwarePublisherCertificateTest/NonSignedAssembly/

```
SoftwarePublisherCertificateTest.NonSignedAssembly

Failing for the past 1 build (Since Unstable#23099 )
Took 5.3 sec.
Error Message

System.Security.Cryptography.CryptographicException was expected
```

**Comment by minego** [comment]
> There is a test failure:
> 
> https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx/23099/testReport/junit/(root)/SoftwarePublisherCertificateTest/NonSignedAssembly/
> 
> ```
> SoftwarePublisherCertificateTest.NonSignedAssembly
> 
> Failing for the past 1 build (Since Unstable#23099 )
> Took 5.3 sec.
> Error Message
> 
> System.Security.Cryptography.CryptographicException was expected
> ```

I don't have any os x systems. I've been looking through code to try to understand the cause, but I think I'm going to need to find a system to debug on before I can learn more.

**Comment by minego** [comment]
> > There is a test failure:
> > https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx/23099/testReport/junit/(root)/SoftwarePublisherCertificateTest/NonSignedAssembly/
> > ```
> > SoftwarePublisherCertificateTest.NonSignedAssembly
> > 
> > Failing for the past 1 build (Since Unstable#23099 )
> > Took 5.3 sec.
> > Error Message
> > 
> > System.Security.Cryptography.CryptographicException was expected
> > ```
> 
> I don't have any os x systems. I've been looking through code to try to understand the cause, but I think I'm going to need to find a system to debug on before I can learn more.

I found a mac build system that I can access via ssh, and I've been able to reproduce this test failure. I hope to find a fix shortly at which point I'll update the pull request.

**Comment by minego** [comment]
I've pushed a fix that fixed the test on the system I have access to.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by akoeplinger** [comment]
@monojenkins commit apidiff

**Comment by akoeplinger** [comment]
@minego Thank you for the contribution!

**Comment by minego** [comment]
Thank you!

**Comment by directhex** [comment]
@minego the private key included in your test case was created expressly for the test case, right? It's not a "real" private key of any kind?

**Comment by minego** [comment]
It was copied from another test that was already in place. If I recall
correctly it was AuthenticodeDeformatterTest.cs

**Pull request merged**